In this notebook, we use the processed data from the `analysis.ipynb` file and applied the various methods on it to find the better r2_score. Saved the final model in pickle format for the prediction and at last done the prediction.

### Importing files/data from analysis.ipynb file for creating model

In [1]:
import pprint
import nltk
import pandas as pd
import pickle
import requests
import numpy as np
import joblib
import import_ipynb

nltk.download('punkt')
pd.set_option('display.max_rows', None)
%run analysis.ipynb import data, store_tag

ModuleNotFoundError: No module named 'import_ipynb'

In [ ]:
data.head()

Description is also play a role in the popularity of your repos, so we need to clean  `description` tuple by removing punctuations and creating word counts.

In [5]:

data['description']=data['description'].str.replace('\W'," ")
data['description']=data['description'].str.lower()

In [6]:
data['most_used_lang']=data['most_used_lang'].str.replace('\W'," ")
data['most_used_lang']=data['most_used_lang'].str.lower()

In [7]:
data['language']=data['most_used_lang'].str.split()
vocab = []
for words in data['language']:
    for word in words:
        vocab.append(word)
        
        
vocab = list(set(vocab))

In [8]:
data.head()

repo_name   star   fork  watch  \
0                                              keras  47900  18100   2100   
1                                       scikit-learn  40300  19600   2200   
2                          PythonDataScienceHandbook  23100   9900   1500   
3  Probabilistic-Programming-and-Bayesian-Methods...  21000   6600   1400   
4                          Data-Science--Cheat-Sheet  18400   8200   1500   

   issue                                               tags  \
0   2940  ['deep-learning', 'tensorflow', 'neural-networ...   
1   1505  ['machine-learning', 'python', 'statistics', '...   
2     65  ['scikit-learn', 'numpy', 'python', 'jupyter-n...   
3    127  ['bayesian-methods', 'pymc', 'mathematical-ana...   
4      7                                                 []   

                                         description  contributers  license  \
0                           deep learning for humans          49.0        1   
1           scikit learn  machine learning in python         108.0        1   
2  python data science handbook  full text in jup...           0.0        1   
3  aka  bayesian methods for hackers   an introdu...           0.0        1   
4                                       cheat sheets           0.0        0   

                                            repo_url    most_used_lang  \
0                https://github.com/keras-team/keras            python   
1       https://github.com/scikit-learn/scikit-learn            python   
2  https://github.com/jakevdp/PythonDataScienceHa...  jupyter notebook   
3  https://github.com/CamDavidsonPilon/Probabilis...  jupyter notebook   
4  https://github.com/abhat222/Data-Science--Chea...       no language   

   tag_ratio             language  
0       9.31             [python]  
1       5.97             [python]  
2       2.42  [jupyter, notebook]  
3       1.72  [jupyter, notebook]  
4       0.00       [no, language]

In [9]:
word_counts_for_language = {unique_word: [0] * len(data['language']) for unique_word in vocab}

for index, words in enumerate(data['language']):
    for word in words:
        word_counts_for_language[word][index] += 1

language_col=pd.DataFrame(word_counts_for_language)

### Feature Engineering

Tokenizing the text data

In [10]:
text=[]
for index,row in data.iterrows():
    tokens=nltk.word_tokenize(row['description'])
    text.append(tokens)

In [11]:
nltk.download('stopwords')
from nltk.corpus import stopwords
stopword=stopwords.words('english')

[nltk_data] Downloading package stopwords to /home/myth/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


To get numerical form of description , we create word dict according to their frequency i.e. the word which occurs more would be have low index value and removing words which has less `freq`(threshold=1)

In [12]:
def txt_to_num(text, threshold_rarewords = 1):
    
    # removing sublist from text list
    if len(text) > 1:
        text_list = [item for sublist in text for item in sublist]
    else:
        text_list = text
    
    # get word freuqncy
    freq_word = nltk.FreqDist(text_list)


    df = pd.DataFrame.from_dict(freq_word, orient='index')
    df.columns = ['freq']
      

    df.sort_values(by=['freq'], ascending=False, inplace=True)
  
    # Add word index
    number_of_words = df.shape[0]
    for word in df.index:
        if word in stopword:
            df.loc[word]['freq']=0
    df.sort_values('freq',inplace=True,ascending=False)
    
    df['word_index'] = list(np.arange(number_of_words)+1)
    
    # indexing 0 for rare words 
    frequency = df['freq'].values
    word_index = df['word_index'].values
    mask = frequency <= threshold_rarewords
    word_index[mask] = 0
    df['word_index'] =  word_index
    
    
    word_dict = df['word_index'].to_dict()
    # print(df.tail())
    # dict for word into num
    text_num = []
    for string in text:
        string_numbers = [word_dict[word] for word in string]
        text_num.append(string_numbers)  
    
    return (text_num, word_dict)

In [13]:
desc_num,word_dict=txt_to_num(text)
txt=[sum(x)  for x in desc_num]
data['desc_to_num']=txt

In [14]:
# repo_text=[]
# for index,row in data.iterrows():
#     tokens=nltk.word_tokenize(row['repo_name'])
#     repo_text.append(tokens)

In [15]:
# repo_num,repo_word_dict=txt_to_num(repo_text)
# txt=[sum(x)  for x in repo_num]
# data['repo_to_num']=txt

In [16]:
data.head()

repo_name   star   fork  watch  \
0                                              keras  47900  18100   2100   
1                                       scikit-learn  40300  19600   2200   
2                          PythonDataScienceHandbook  23100   9900   1500   
3  Probabilistic-Programming-and-Bayesian-Methods...  21000   6600   1400   
4                          Data-Science--Cheat-Sheet  18400   8200   1500   

   issue                                               tags  \
0   2940  ['deep-learning', 'tensorflow', 'neural-networ...   
1   1505  ['machine-learning', 'python', 'statistics', '...   
2     65  ['scikit-learn', 'numpy', 'python', 'jupyter-n...   
3    127  ['bayesian-methods', 'pymc', 'mathematical-ana...   
4      7                                                 []   

                                         description  contributers  license  \
0                           deep learning for humans          49.0        1   
1           scikit learn  machine learning in python         108.0        1   
2  python data science handbook  full text in jup...           0.0        1   
3  aka  bayesian methods for hackers   an introdu...           0.0        1   
4                                       cheat sheets           0.0        0   

                                            repo_url    most_used_lang  \
0                https://github.com/keras-team/keras            python   
1       https://github.com/scikit-learn/scikit-learn            python   
2  https://github.com/jakevdp/PythonDataScienceHa...  jupyter notebook   
3  https://github.com/CamDavidsonPilon/Probabilis...  jupyter notebook   
4  https://github.com/abhat222/Data-Science--Chea...       no language   

   tag_ratio             language  desc_to_num  
0       9.31             [python]          737  
1       5.97             [python]          477  
2       2.42  [jupyter, notebook]         1116  
3       1.72  [jupyter, notebook]         9459  
4       0.00       [no, language]         1439

From above we can see that their are some numerical values and stopwords too, so we need to remove them from vocabulary list

In [17]:
training_set = pd.concat([data, language_col],axis=1)
training_set.columns

Index(['repo_name', 'star', 'fork', 'watch', 'issue', 'tags', 'description',
       'contributers', 'license', 'repo_url', 'most_used_lang', 'tag_ratio',
       'language', 'desc_to_num', 'purebasic', 'emacs', 'jupyter', 'no',
       'assembly', 'matlab', 'viml', 'kotlin', 'groff', 'groovy', 'vue',
       'scala', 'erlang', 'lisp', 'java', 'typescript', 'actionscript',
       'shaderlab', 'scheme', 'ruby', 'pascal', 'swift', 'javascript',
       'coffeescript', 'jsonnet', 'clojure', 'r', 'apex', 'objective', 'dart',
       'html', 'script', 'arc', 'rust', 'lua', 'qml', 'php', 'dockerfile',
       'perl', 'notebook', 'vhdl', 'powershell', 'vim', 'tex', 'c', 'pure',
       'makefile', 'eagle', 'julia', 'css', 'data', 'shell', 'ragel', 'elixir',
       'ocaml', 'racket', 'processing', 'go', 'language', 'python', 'roff',
       'xslt', 'freemarker', 'io', 'other', 'hcl', 'haskell'],
      dtype='object')

In [18]:
training_set.drop(['repo_name', 'tags', 'description', 'repo_url', 'language', 'no', 'most_used_lang'], axis=1, inplace=True)

In [19]:
training_set = training_set.dropna()

### Training model

In [20]:
## LR can be used:
# gradient descent
# random forest 
# NN regression (Relu)
# Lasso regression
# Ridge regression
# Elastic net regression

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
X_train, X_test, y_train, y_test = train_test_split(training_set.iloc[:,1:], training_set.iloc[:,0], test_size=0.2, random_state=22)

In [22]:
X_train.shape

(1200, 72)

In [23]:
ac=0

Gradient Descent

In [24]:
from sklearn.linear_model import SGDRegressor

sgd=SGDRegressor(loss='squared_loss',alpha=0.3,max_iter=5000)
sgd.fit(X_train, y_train)

predictions = sgd.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print('Mean Squared Error:', mse)
accu=r2_score(y_test, predictions)
print("R2 score: ", accu)

if accu > ac:
    model = sgd
    ac = accu

3.303941592424989e+16
R2 score:  -4.2170038614981604e+24


Normal Equation

In [25]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(X_train, y_train)

predictions = lr.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print("Mean Squared Error: ", mse)
accu=r2_score(y_test, predictions)
print("R2 score: ",accu)

if accu > ac:
    model = lr
    ac = accu

5681.986682232555
R2 score:  0.8752789669984791


Random forest

In [26]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(max_depth=8, random_state=25)
rf.fit(X_train, y_train)

predictions = rf.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print("Mean Squared Error: ", mse)
accu=r2_score(y_test, predictions)
print("R2 score: ",accu)

if accu > ac:
    model = rf
    ac = accu

4801.792011241244
R2 score:  0.9109270233630825


Lasso Regression


In [27]:
from sklearn import linear_model

las = linear_model.Lasso(alpha=0.3,max_iter=3000)
las.fit(X_train, y_train)

predictions = las.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print("Mean Squared Error: ", mse)
accu=r2_score(y_test, predictions)
print("R2 score: ",accu)

if accu > ac:
    model = las
    ac = accu

5680.798186828632
R2 score:  0.8753311370862266


Ridge Regression

In [28]:
from sklearn import linear_model

las = linear_model.Ridge(alpha=0.3,max_iter=3000)
las.fit(X_train, y_train)

predictions = las.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print("Mean Squared Error: ", mse)
accu=r2_score(y_test, predictions)
print("R2 score: ",accu)

if accu > ac:
    model = las
    ac = accu

5689.720420421575
R2 score:  0.874939220955455


Elastic Net Regression

In [29]:
from sklearn.linear_model import ElasticNet

elsnr = ElasticNet(alpha=0.03,max_iter=2000)
elsnr.fit(X_train, y_train)

predictions = elsnr.predict(X_test)
mse = mean_squared_error(y_test, predictions, squared=False)

print("Mean Squared Error: ", mse)
accu=r2_score(y_test, predictions)
print("R2 score: ",accu)

if accu > ac:
    model = elsnr
    ac = accu

5848.999716530464
R2 score:  0.8678392553627553


All results:

| Methods                | MSE      | R2 Score |
|------------------------|----------|----------|
| Gradient Descent         | 3.303    | -4.217 |
| Linear Regression        | 5681.986 | 0.875  |
| Random Forest            | 4801.79  | 0.910  |
| Lasso Regression         | 5680.798 | 0.8753 |
| Ridge Regression         | 5689.72  | 0.8749 |
| Elastic Net Regression   | 5848.99  | 0.868  |

In the above table, we can see that the `Random Forest` perform better than others methods with 0.91 r2 score.

In [30]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV
from numpy import absolute

In [29]:
# xg = XGBRegressor(verbosity=1, eta=0.2)
# params = {'min_child_weight':[4,5], 'gamma':[i/10.0 for i in range(3,6)],  'subsample':[i/10.0 for i in range(6,11)],
# 'colsample_bytree':[i/10.0 for i in range(6,11)], 'max_depth': [4,6,7]}
# gb = GridSearchCV(xg,params)
# gb.fit(X_train, y_train)

# predictions = gb.predict(X_test)
# mse = mean_squared_error(y_test, predictions, squared=False)

# print("Mean Squared Error: ", mse)
# accu=r2_score(y_test, predictions)
# print("R2 score: ", accu)

# if accu > ac:
#     model = gb
#     ac = accu

In [31]:
# Saving the model in `pickle` format
with open('../pretrained_model/train_model.pkl', 'wb') as f:  
    pickle.dump(model, f)
with open('../pretrained_model/objs.pkl', 'wb') as f:  
    pickle.dump([store_tag, vocabulary, vocab, word_dict], f)

Testing the model on the repo to predict its popularity.

In [32]:
user_name=input("Enter username:")
repo_name=input("Enter repo name:")

In [36]:
header={'Accept':'application/vnd.github.mercy-preview+json',
'visibility':'PUBLIC',
"Authorization": "token PASTE_YOUR_GITHUB_TOKEN",
}

request=requests.get('https://api.github.com/repos/'+user_name+'/'+repo_name,headers=header)
data=request.json()
keys=['forks','watchers','open_issues_count','topics','description','language']

dataf=[data[k] for k in keys]
if str(data['license']).lower()=='none' or data['license']:
    license=0
else:
    license=1

request=requests.get('https://api.github.com/repos/'+user_name+'/'+repo_name+'contributors',headers=header)
dat=request.json()
contributors=len(dat)
dataf.append(contributors)
dataf.append(license)
df=pd.DataFrame([dataf],columns=['fork','watch','issue','tags','description','most_used_lang','contributers','license'])
def perc(tags):
    sum_of_perc=0
    for tag in tags:
        if tag in store_tag:
            sum_of_perc+=(store_tag[tag]/vocabulary)

    return (sum_of_perc*100)

df['tag_ratio']=df['tags'].apply(perc).astype(float)
text=nltk.word_tokenize(df['description'][0])
tex=nltk.word_tokenize(repo_name)

def string_num(text):
    string_numbers=0
    for string in text:
        if string in word_dict:
            string_numbers+=word_dict[string]
    return string_numbers

df['desc_to_num']=string_num(text)

word_counts_for_language = {unique_word: [0] for unique_word in vocab}

df['most_used_lang']=df['most_used_lang'].str.split()
for word in df['most_used_lang']:
    if word[0] in word_counts_for_language:
        word_counts_for_language[word[0].lower()][0] += 1

language_col=pd.DataFrame(word_counts_for_language)
test_set=pd.concat([df,language_col],axis=1)
test_set.drop(['tags','description','language','no','most_used_lang'], axis=1, inplace=True)

print("Your repo Popularity:",int(model.predict(test_set)[0]))

your repo Popularity: 35
